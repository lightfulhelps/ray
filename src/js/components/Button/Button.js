// @flow
import * as React from 'react';
import { color, number } from '@storybook/addon-knobs';
import classNames from 'classnames';
import Icon from '../Icon/Icon';

type Props = {
  children: React.Element<any>,
  className?: string,
  isBlock?: boolean,
  isDisabled?: boolean,
  isOutline?: boolean,
  onClick?: () => void,
  size?: 'lg' | 'md' | 'sm',
  tag?: string,
  theme?: string,
  icon?: string,
  iconPosition?: string,
  isDropdown?: boolean,
};

const Button = ({
  children,
  label,
  className,
  theme = 'primary',
  isBlock,
  isDisabled,
  isOutline,
  onClick = () => {},
  size,
  tag: Tag = 'button',
  icon,
  iconPosition,
  isDropdown,
}: Props) => {
  const classes = classNames(
    className,
    'btn',
    `btn${isOutline ? '-outline' : ''}-${theme}`,
    size ? `btn-${size}` : false,
    { 'btn-block': isBlock },
    { disabled: isDisabled },
    { 'btn-icon': icon && !label},
    { [`btn-icon-${iconPosition}`]: iconPosition && label }
  );

  return (
    <Tag 
      className={classes} 
      onClick={onClick}
      data-toggle={isDropdown && "dropdown"}
      aria-haspopup={isDropdown && "true"}
      aria-expanded={isDropdown && "false"}
    >
      {label && (
        <span>{label}</span>
      )}
      {icon && (
        <Icon
          name={icon}
          size={number('Size', 24)}
          color={color('Color', '#adb5bd')}
          hoverColor={color('Hover Color', '#212529')}
        />
      )}
    </Tag>
  );
};

export default Button;
