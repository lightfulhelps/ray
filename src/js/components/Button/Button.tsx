import * as React from 'react';
import classNames from 'classnames';
import Icon from '../Icon/Icon';
import { IconNameType } from '../Icon/icons';

// Size is an attribute of the input html element so there was a conflict with our size props.
// Hence the need to omit this attribute
type ButtonAttributes = Omit<React.HTMLProps<HTMLButtonElement>, 'size'>;

export type Props = {
  animateGradient?: boolean;
  children?: React.ReactNode;
  className?: string;
  forceSolidColor?: boolean;
  icon?: IconNameType;
  iconPosition?: 'left' | 'right';
  iconSize?: string;
  iconTheme?: string;
  isBlock?: boolean;
  isDisabled?: boolean;
  isOutline?: boolean;
  loading?: boolean;
  onClick?: (event: React.MouseEvent) => void;
  size?: 'lg' | 'md' | 'sm';
  tag?: keyof JSX.IntrinsicElements | typeof React.Component;
  theme?: string;
  type?: 'button' | 'submit' | 'reset';
  [key: string]: any;
};

const gradientThemes = [
  'primary',
  'secondary',
  'tertiary-red',
  'tertiary-green',
  'tertiary-black',
  'academy',
  'success',
  'neutral',
];
const Button: React.FC<ButtonAttributes & Props> = ({
  children,
  className,
  forceSolidColor = false,
  icon,
  iconPosition = 'left',
  iconSize = '1em',
  iconTheme,
  isBlock,
  isDisabled,
  isOutline,
  loading = false,
  onClick,
  size,
  tag: Tag = 'button',
  theme = 'primary',
  type = 'button',
  animateGradient = false,
  ...other
}: Props) => {
  const classes = classNames(
    className,
    'btn',
    { [`btn${isOutline ? '-outline' : ''}-gradient-${theme}`]: !forceSolidColor },
    `btn${isOutline ? '-outline' : ''}-${theme}`, // backup
    size ? `btn-${size}` : false,
    { disabled: isDisabled },
    { 'btn-icon': icon && !children },
    { [`btn-icon-${iconPosition}`]: icon && children },
    { loading },
    { 'animate-gradient': animateGradient }
  );

  const handleClick = (e: React.MouseEvent): void => {
    if (isDisabled) {
      e.preventDefault();
      return;
    }

    if (typeof onClick === 'function') onClick(e);
  };

  // Workaround to optionally add type attribute since this one is only available on button Jsx element
  const optionalProps: { [key: string]: any } = {};
  optionalProps.type = Tag === 'button' ? type : undefined;

  const buttonComponent = (
    <Tag
      {...other}
      {...optionalProps}
      className={classes}
      onClick={handleClick}
      disabled={isDisabled}
    >
      <span>
        {children}
        {icon && (
          <Icon
            name={icon}
            size={iconSize}
            theme={iconTheme || (isOutline && theme)}
            isGradient={gradientThemes.includes(theme) && !forceSolidColor}
          />
        )}
      </span>
    </Tag>
  );

  if (isBlock) return <div className="d-grid">{buttonComponent}</div>;
  return buttonComponent;
};

export default Button;
