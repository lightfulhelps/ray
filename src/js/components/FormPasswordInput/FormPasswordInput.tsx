import React, { useState } from 'react';
import classNames from 'classnames';
import { FormInput, Button } from '../..';

interface Props extends React.HTMLAttributes<HTMLInputElement> {
  className?: string;
  id?: string;
  isInvalid?: boolean;
  isValid?: boolean;
  placeholder?: string;
  value?: string;
  size?: 'sm' | 'md' | 'lg';
  name?: string;
}

const FormPasswordInput: React.FC<Props> = ({
  className,
  onChange,
  onBlur,
  value,
  id,
  isInvalid,
  isValid,
  name = 'password',
  placeholder,
  size,
  ...other
}) => {
  const [hidePassword, setHidePassword] = useState(false);

  const togglePassword = () => {
    setHidePassword(!hidePassword);
  };

  const classes = classNames(className, 'form-control');

  return (
    <div className="form-password-input input-group">
      <FormInput
        {...other}
        className={classes}
        data-test-id="password-input"
        isInvalid={isInvalid}
        isValid={isValid}
        name={name}
        id={id}
        onBlur={onBlur}
        onChange={onChange}
        placeholder={placeholder}
        size={size}
        type={hidePassword ? 'password' : 'text'}
        value={value}
      />
      <div className="input-group-append">
        <Button
          data-test-id="password-input-button"
          icon={hidePassword ? 'preview' : 'previewHide'}
          isOutline
          onClick={togglePassword}
          size={size}
          theme="gray-600"
        >
          <span className="form-password-input__toggle-password-text">
            {hidePassword ? 'Show' : 'Hide'}
          </span>
        </Button>
      </div>
    </div>
  );
};

export default FormPasswordInput;
