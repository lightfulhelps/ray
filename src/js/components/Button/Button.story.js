import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, select, boolean } from '@storybook/addon-knobs';
import Button from './Button';

import '../../../scss/ray.scss';

const stories = storiesOf('Button', module);

const themes = {
  primary: 'Primary',
  secondary: 'Secondary',
  success: 'Success',
  danger: 'Danger',
  warning: 'Warning',
  info: 'Info',
  light: 'Light',
  dark: 'Dark',
};

const sizes = {
  lg: 'Large',
  md: 'Medium',
  sm: 'Small',
};

stories.addDecorator(withKnobs);

stories.add('Default', () => (
  <Button
    onClick={action('clicked')}
    color={select('Theme', themes, 'primary')}
    outline={boolean('Outline', false)}
    disabled={boolean('Disabled', false)}
    size={select('Size', sizes, 'md')}
  >
    {text('Text', 'Click Me')}
  </Button>
));
