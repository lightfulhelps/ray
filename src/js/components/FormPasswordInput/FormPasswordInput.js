// @flow
import * as React from 'react';
import classNames from 'classnames';
import Button from '../Button/Button';

type Props = {
  className?: string,
};

type State = {
  hidePassword: boolean,
};

class FormPasswordInput extends React.Component<Props, State> {
  state = {
    hidePassword: true,
  };

  togglePassword = () => {
    this.setState(prevState => ({
      hidePassword: !prevState.hidePassword,
    }));
  };

  render() {
    const { className, ...other } = this.props;
    const { hidePassword } = this.state;
    const classes = classNames(className, 'form-password-input', 'input-group');

    const inputType = hidePassword ? 'password' : 'text';
    const buttonText = hidePassword ? 'Show' : 'Hide';
    const icon = hidePassword ? 'preview' : 'previewHide';

    return (
      <div className={classes}>
        <input type={inputType} className="form-control" {...other} />
        <div className="input-group-append">
          <Button onClick={this.togglePassword} icon={icon} isOutline theme="gray-600">
            <span className="form-password-input__toggle-password-text">{buttonText}</span>
          </Button>
        </div>
      </div>
    );
  }
}

export default FormPasswordInput;
