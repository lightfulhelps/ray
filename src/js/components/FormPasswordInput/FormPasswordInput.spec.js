import React from 'react';
import { shallow } from 'enzyme';
import merge from 'lodash/merge';
import FormPasswordInput from './FormPasswordInput';
import Button from '../Button/Button';

const setup = (overrides = {}) => {
  const props = merge(
    {
      onBlur: jest.fn(),
      onChange: jest.fn(),
      value: 'Some value',
    },
    overrides
  );
  const wrapper = shallow(<FormPasswordInput {...props} />);

  return { wrapper, props, button: wrapper.find(Button), input: wrapper.find('input') };
};

describe('<FormPasswordInput />', () => {
  it('should render', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle className', () => {
    const { wrapper } = setup();

    wrapper.setProps({ className: 'custom' });

    expect(wrapper.hasClass('form-password-input')).toBe(true);
    expect(wrapper.hasClass('custom')).toBe(true);
  });

  it('should pass through other props', () => {
    const { wrapper } = setup({ tabIndex: 1, 'data-test': 'test', className: 'custom' });

    expect(wrapper.prop('tabIndex')).toEqual(1);
    expect(wrapper.prop('data-test')).toEqual('test');
  });

  it('defaults name parameter to password', () => {
    const { input } = setup();

    expect(input.prop('name')).toEqual('password');
  });

  describe('input pass throughs', () => {
    it('passes through placeholder', () => {
      const placeholder = 'Some placeholder';

      const { input } = setup({ placeholder });

      expect(input.prop('placeholder')).toEqual(placeholder);
    });

    it('passes through value', () => {
      const value = 'Some value';

      const { input } = setup({ value });

      expect(input.prop('value')).toEqual(value);
    });

    it('passes through onChange', () => {
      const {
        input,
        props: { onChange },
      } = setup();

      input.simulate('change');

      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('passes through onBlur', () => {
      const {
        input,
        props: { onBlur },
      } = setup();

      input.simulate('blur');

      expect(onBlur).toHaveBeenCalledTimes(1);
    });

    it('passes through name', () => {
      const name = 'secretPassword';
      const { input } = setup({ name });

      expect(input.prop('name')).toEqual(name);
    });

    it('passes through id', () => {
      const id = 'secretPassword';
      const { input } = setup({ id });

      expect(input.prop('id')).toEqual(id);
    });
  });

  describe('end to end test helpers', () => {
    it('has a data-test-id for the input', () => {
      const { input } = setup();

      expect(input.prop('data-test-id')).toEqual('password-input');
    });

    it('has a data-test-id for the button', () => {
      const { button } = setup();

      expect(button.prop('data-test-id')).toEqual('password-input-button');
    });
  });

  describe('show password toggle', () => {
    it('hides password by default', () => {
      const { input } = setup();

      expect(input.prop('type')).toEqual('password');
    });

    it('shows password when state is set', () => {
      const { wrapper } = setup();

      wrapper.setState({ hidePassword: false });

      const input = wrapper.find('input');

      expect(input.prop('type')).toEqual('text');
    });

    describe('icon toggle', () => {
      it('has button for toggle', () => {
        const { button } = setup();

        expect(button).toHaveLength(1);
      });

      it('has a button that changes state when clicked', () => {
        const { button, wrapper } = setup();

        button.simulate('click');

        expect(wrapper.state('hidePassword')).toEqual(false);
      });

      it('has a button that says Show when password is hidden', () => {
        const { wrapper } = setup();

        const span = wrapper.find('.form-password-input__toggle-password-text');

        expect(span.text()).toEqual('Show');
      });

      it('has a button that says Hide when password is hidden', () => {
        const { wrapper } = setup();

        wrapper.setState({ hidePassword: false });

        const span = wrapper.find('.form-password-input__toggle-password-text');

        expect(span.text()).toEqual('Hide');
      });

      it('has a preview icon when password is hidden', () => {
        const { button, wrapper } = setup();

        expect(wrapper.state('hidePassword')).toEqual(true);

        expect(button.prop('icon')).toEqual('preview');
      });

      it('has a preview-hide icon when password is hidden', () => {
        const { wrapper } = setup();

        wrapper.setState({ hidePassword: false });

        expect(wrapper.find(Button).prop('icon')).toEqual('previewHide');
      });
    });
  });
});
