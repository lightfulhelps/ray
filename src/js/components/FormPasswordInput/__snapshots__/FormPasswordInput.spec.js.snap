// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormPasswordInput /> should render 1`] = `
<div
  className="form-password-input input-group"
>
  <input
    className="form-control"
    data-test-id="password-input"
    name="password"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    type="password"
    value="Some value"
  />
  <div
    className="input-group-append"
  >
    <Button
      data-test-id="password-input-button"
      icon="preview"
      isOutline={true}
      onClick={[Function]}
      theme="gray-600"
    >
      <span
        className="form-password-input__toggle-password-text"
      >
        Show
      </span>
    </Button>
  </div>
</div>
`;
